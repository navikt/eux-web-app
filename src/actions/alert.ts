import * as types from 'constants/actionTypes'
import { ActionWithPayload } from '@navikt/fetch'
import { Action, ActionCreator } from 'redux'

export const alertClear: ActionCreator<Action> = (): Action => ({
  type: types.ALERT_CLEAR
})

export const alertFailure: ActionCreator<ActionWithPayload<{message: JSX.Element | string}>> = (
  message: JSX.Element | string
): ActionWithPayload<{message: JSX.Element | string}> => ({
  type: types.ALERT_FAILURE,
  payload: {
    message
  }
})

export const alertSuccess: ActionCreator<ActionWithPayload<{message: JSX.Element | string}>> = (
  message: JSX.Element | string
): ActionWithPayload<{message: JSX.Element | string}> => ({
  type: types.ALERT_SUCCESS,
  payload: {
    message
  }
})

export const alertWarning: ActionCreator<ActionWithPayload<{message: JSX.Element | string}>> = (
  message: JSX.Element | string
): ActionWithPayload<{message: JSX.Element | string}> => ({
  type: types.ALERT_WARNING,
  payload: {
    message
  }
})
