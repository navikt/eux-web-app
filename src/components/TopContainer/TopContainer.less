@import '../../constants';

body {
  background-color: @white;
  display: flex;
  flex-direction: column;
  min-height: 100vh;

  main {
    flex: 1 0 auto;
    display: flex;
    flex-direction: column;
    min-height: 100vh;
  }

  .highContrast {
      color: @navOransje !important;
      background-color: @black !important;
  }
}
