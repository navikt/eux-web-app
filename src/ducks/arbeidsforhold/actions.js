
/**
 * Actions
 * -----------------------------------------------------------------------
 * Dette er action creators som returnerer Redux-klargjorte actions
 * uten support for asynkrone kall.
 */
