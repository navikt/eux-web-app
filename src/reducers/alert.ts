import * as types from 'constants/actionTypes'
import { ActionWithPayload } from 'js-fetch-api'
import _ from 'lodash'
import { Action } from 'redux'

export interface AlertState {
  clientErrorStatus: string | undefined;
  clientErrorMessage: string | undefined;
  serverErrorMessage: string | undefined;
  uuid: string | undefined;
  error: any | undefined;
  type: string | undefined;
}

export const initialAlertState: AlertState = {
  clientErrorStatus: undefined,
  clientErrorMessage: undefined,
  serverErrorMessage: undefined,
  uuid: undefined,
  error: undefined,
  type: undefined
}

const alertReducer = (state: AlertState = initialAlertState, action: Action | ActionWithPayload = { type: '' }): AlertState => {
  let clientErrorMessage: string | undefined, serverErrorMessage: string, clientErrorStatus: string

  if (
    action.type === types.ALERT_CLIENT_CLEAR ||
    action.type === types.APP_CLEAN_DATA ||
    action.type === types.SAK_PERSON_RELATERT_RESET ||
    action.type === types.SAK_PERSON_GET_REQUEST ||
    action.type === types.SAK_PERSON_RELATERT_SEARCH_REQUEST ||
    action.type === types.SAK_ABROADPERSON_ADD_SUCCESS ||
    action.type === types.SAK_TPSPERSON_ADD_SUCCESS) {
    return initialAlertState
  }

  if (_.endsWith(action.type, '/ERROR')) {
    switch (action.type) {
      case types.SERVER_INTERNAL_ERROR:
        serverErrorMessage = 'message:error-serverInternalError'
        break

      case types.SERVER_UNAUTHORIZED_ERROR:
        serverErrorMessage = 'message:error-serverAuthenticationError'
        break

      default:
        serverErrorMessage = (action as ActionWithPayload).payload.message || 'message:error-serverInternalError'
        break
    }

    return {
      ...state,
      type: action.type,
      serverErrorMessage: serverErrorMessage,
      error: (action as ActionWithPayload).payload ? (action as ActionWithPayload).payload.error : undefined,
      uuid: (action as ActionWithPayload).payload ? (action as ActionWithPayload).payload.uuid : undefined
    }
  }

  if (_.endsWith(action.type, '/FAILURE')) {
    clientErrorStatus = 'ERROR'

    switch (action.type) {
      case types.SAK_PERSON_GET_FAILURE:
        clientErrorMessage = 'message:error-person-notFound'
        break

      case types.SAK_PERSON_RELATERT_SEARCH_FAILURE:
        clientErrorMessage = 'message:error-personRelated-notFound'
        break

      case types.SAK_ABROADPERSON_ADD_FAILURE:
        clientErrorMessage = 'message:error-abroadperson-exists'
        break

      case types.SAK_TPSPERSON_ADD_FAILURE:
        clientErrorMessage = 'message:error-tpsperson-exists'
        break

      case types.SVARPASED_SED_CREATE_FAILURE:
        if ((action as ActionWithPayload).status === 409) {
          clientErrorMessage = 'message:error-svarPaSed-failure-duplicate'
        } else {
          clientErrorMessage = 'message:error-svarPaSed-failure'
        }
        break

      default:
        if ((action as ActionWithPayload).payload && (action as ActionWithPayload).payload.error) {
          clientErrorMessage = (action as ActionWithPayload).payload.error
        } else {
          clientErrorMessage = 'ui:error'
        }
        break
    }

    return {
      ...state,
      type: action.type,
      clientErrorStatus: clientErrorMessage ? clientErrorStatus : undefined,
      clientErrorMessage: clientErrorMessage,
      error: (action as ActionWithPayload).payload ? (action as ActionWithPayload).payload.error : undefined,
      uuid: (action as ActionWithPayload).payload ? (action as ActionWithPayload).payload.uuid : undefined
    }
  }

  if (action.type === types.LOCALSTORAGE_ENTRY_SAVE) {
    clientErrorMessage = 'message:success-svarPaSed-localstorage-save'
  }
  if (action.type === types.SVARPASED_REPLYSED_SET) {
    clientErrorMessage = undefined
  }

  if (!clientErrorMessage) {
    return state
  }

  return {
    ...state,
    type: action.type,
    clientErrorStatus: 'OK',
    clientErrorMessage: clientErrorMessage,
    uuid: undefined,
    error: undefined
  }
}

export default alertReducer
