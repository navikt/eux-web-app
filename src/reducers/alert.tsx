import { ExternalLink } from '@navikt/ds-icons'
import * as types from 'constants/actionTypes'
import i18n from 'i18n'
import { ActionWithPayload } from 'js-fetch-api'
import _ from 'lodash'
import { FlexDiv, HorizontalSeparatorDiv } from 'nav-hoykontrast'
import { Link } from '@navikt/ds-react'
import React from 'react'
import { Action } from 'redux'

export interface AlertState {
  stripeStatus: string | undefined
  stripeMessage: JSX.Element | string | undefined
  bannerStatus: string | undefined
  bannerMessage: string | undefined
  uuid: string | undefined
  error: any | undefined
  type: string | undefined
}

export const initialAlertState: AlertState = {
  stripeStatus: undefined,
  stripeMessage: undefined,
  bannerStatus: undefined,
  bannerMessage: undefined,
  uuid: undefined,
  error: undefined,
  type: undefined
}

const alertReducer = (state: AlertState = initialAlertState, action: Action | ActionWithPayload = { type: '' }): AlertState => {
  let stripeMessage: JSX.Element | string | undefined
  let bannerMessage: string | undefined
  let stripeStatus: string
  let bannerStatus: string

  if (
    action.type === types.ALERT_CLIENT_CLEAR ||
    action.type === types.APP_CLEAN_DATA ||
    action.type === types.SAK_PERSON_RELATERT_RESET ||
    action.type === types.SAK_PERSON_GET_REQUEST ||
    action.type === types.SAK_PERSON_RELATERT_SEARCH_REQUEST ||
    action.type === types.SAK_ABROADPERSON_ADD_SUCCESS ||
    action.type === types.SAK_TPSPERSON_ADD_SUCCESS) {
    return initialAlertState
  }

  /**
   * All ERROR MESSAGES go here, to banner alert
   */
  if (_.endsWith(action.type, '/ERROR')) {
    bannerStatus = 'error'
    switch (action.type) {
      case types.SERVER_INTERNAL_ERROR:
        bannerMessage = i18n.t('message:error-serverInternalError')
        break

      case types.SERVER_UNAUTHORIZED_ERROR:
        bannerMessage = i18n.t('message:error-serverAuthenticationError')
        break

      default:
        bannerMessage = (action as ActionWithPayload).payload.message || i18n.t('message:error-serverInternalError')
        break
    }

    return {
      ...state,
      type: action.type,
      bannerMessage: bannerMessage,
      bannerStatus: bannerStatus,
      error: (action as ActionWithPayload).payload
        ? _.isString((action as ActionWithPayload).payload.error)
            ? (action as ActionWithPayload).payload.error
            : (action as ActionWithPayload).payload.error?.message
        : undefined,
      uuid: (action as ActionWithPayload).payload ? (action as ActionWithPayload).payload.uuid : undefined
    }
  }

  /**
   * All FAILURE MESSAGES go here, to stripe alert
   */
  if (_.endsWith(action.type, '/FAILURE')) {
    stripeStatus = 'error'
    switch (action.type) {
      case types.SAK_PERSON_GET_FAILURE:
        stripeMessage = i18n.t('message:error-person-notFound')
        break

      case types.SAK_PERSON_RELATERT_SEARCH_FAILURE:
        stripeMessage = i18n.t('message:error-personRelated-notFound')
        break

      case types.SAK_ABROADPERSON_ADD_FAILURE:
        stripeMessage = i18n.t('message:error-abroadperson-exists')
        break

      case types.SAK_TPSPERSON_ADD_FAILURE:
        stripeMessage = i18n.t('message:error-tpsperson-exists')
        break

      case types.SVARSED_SED_CREATE_FAILURE:
        if ((action as ActionWithPayload).status === 409) {
          const url = (action as ActionWithPayload).context.sakUrl
          stripeMessage = (
            <FlexDiv>
              <span>{i18n.t('message:error-svarsed-failure-duplicate')}</span>
              <HorizontalSeparatorDiv size='0.5' />
              <Link target='_blank' href={url} rel='noreferrer'>
                <span>
                  {i18n.t('message:error-svarsed-failure-duplicate-2')}
                </span>
                <HorizontalSeparatorDiv size='0.35' />
                <ExternalLink />
              </Link>
            </FlexDiv>
          )
        } else {
          stripeMessage = i18n.t('message:error-svarsed-failure')
        }
        break

      default:
        if ((action as ActionWithPayload).payload && (action as ActionWithPayload).payload.error) {
          stripeMessage = _.isString((action as ActionWithPayload).payload.error)
            ? (action as ActionWithPayload).payload.error
            : (action as ActionWithPayload).payload.error?.message
        } else {
          stripeMessage = i18n.t('ui:error')
        }
        break
    }

    return {
      ...state,
      type: action.type,
      stripeStatus: stripeMessage ? stripeStatus : undefined,
      stripeMessage: stripeMessage,
      error: (action as ActionWithPayload).payload
        ? _.isString((action as ActionWithPayload).payload.error)
            ? (action as ActionWithPayload).payload.error
            : (action as ActionWithPayload).payload.error?.message
        : undefined,
      uuid: (action as ActionWithPayload).payload ? (action as ActionWithPayload).payload.uuid : undefined
    }
  }

  /**
   * All OK MESSAGES for banner go here
   */
  let dealWithBanner = false
  if (action.type === types.APP_CLIPBOARD_COPY) {
    bannerMessage = i18n.t('message:success-clipboard-copy')
    dealWithBanner = true
  }

  if (action.type === types.LOCALSTORAGE_ENTRY_SAVE) {
    bannerMessage = i18n.t('message:success-localstorage-save')
    dealWithBanner = true
  }

  if (dealWithBanner) {
    return {
      ...state,
      type: action.type,
      bannerStatus: 'success',
      bannerMessage: bannerMessage,
      uuid: undefined,
      error: undefined
    }
  }

  /**
   * All OK MESSAGES for stripe go here
   */

  stripeStatus = 'success'

  if (action.type === types.SVARSED_REPLYSED_SET) {
    stripeMessage = undefined
  }

  if (!stripeMessage) {
    return state
  }

  return {
    ...state,
    type: action.type,
    stripeStatus: stripeStatus,
    stripeMessage: stripeMessage,
    uuid: undefined,
    error: undefined
  }
}

export default alertReducer
