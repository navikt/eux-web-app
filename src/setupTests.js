/* eslint-disable */

import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import each from 'jest-each';

global.fetch = require('jest-fetch-mock');

Enzyme.configure({ adapter: new Adapter() });

global.shallow = shallow;
global.mount = mount;

global.each = each;
