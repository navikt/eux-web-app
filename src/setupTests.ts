import {render, screen} from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import { useDispatch, useSelector } from 'react-redux'
import 'jest-styled-components'
import '@testing-library/jest-dom/extend-expect'

jest.mock('react-redux');

(global as any).screen = screen;
(global as any).render = render;
(global as any).act = act

// eslint-disable-next-line no-undef
HTMLCanvasElement.prototype.getContext = jest.fn()
window.scrollTo = jest.fn()

Object.defineProperty(window, 'getComputedStyle', {
  value: () => ({
    getPropertyValue: () => ('')
  })
})

jest.mock('amplitude-js', () => ({
  getInstance: () => ({
    init: jest.fn(),
    logEvent: jest.fn()
  })
}))

jest.mock('i18next', () => {
  const use = jest.fn()
  const init = jest.fn()
  const loadLanguages = jest.fn()
  const result = {
    use: use,
    init: init,
    loadLanguages: loadLanguages
  }
  use.mockImplementation(() => result)
  return result
})

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: any) => key })
}))

export const stageSelector = (defaultSelector: any, params: any) => {
  (useDispatch as jest.Mock).mockImplementation(() => jest.fn());
  (useSelector as jest.Mock).mockImplementation(() => ({
    ...defaultSelector,
    ...params
  }))
}
