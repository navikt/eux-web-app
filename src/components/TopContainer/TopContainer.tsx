import { clientClear, clientError } from 'actions/alert'
import { closeModal } from 'actions/ui'
import classNames from 'classnames'
import Alert from 'components/Alert/Alert'
import Header from 'components/Header/Header'
import Modal from 'components/Modal/Modal'
import SessionMonitor from 'components/SessionMonitor/SessionMonitor'
import Version from 'components/Version/Version'
import { ModalContent } from 'declarations/components'
import { State } from 'declarations/reducers'
import _ from 'lodash'
import NavHighContrast, {
  fadeIn,
  PileDiv, slideInFromBottom,
  slideInFromLeft,
  slideInFromRight,
  slideInFromTop,
  themeKeys
} from 'nav-hoykontrast'
import Error from 'pages/Error'
import PT from 'prop-types'
import React from 'react'
import { ErrorBoundary } from 'react-error-boundary'
import { useTranslation } from 'react-i18next'
import { useDispatch, useSelector } from 'react-redux'
import styled, { createGlobalStyle } from 'styled-components'

const GlobalStyle = createGlobalStyle`
body {
  margin: 0;
  padding: 0;
  color: ${({ theme } : any) => theme[themeKeys.MAIN_FONT_COLOR]};
  background: ${({ theme } : any) => theme[themeKeys.MAIN_BACKGROUND_COLOR]};
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  font-family: 'Source Sans Pro', Arial, sans-serif;
}
#root {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}
.nolabel {
  margin-top: 1.8rem;
}
.fadeIn {
  opacity: 0;
  animation: ${fadeIn} 0.25s forwards;
}
.slideInFromLeft {
  opacity: 0;
  transform: translateX(-20px);
  animation: ${slideInFromLeft(20)} 0.3s forwards;
}
.slideInFromRight {
  opacity: 0;
  transform: translateX(20px);
  animation: ${slideInFromRight(20)} 0.3s forwards;
}
.slideInFromTop {
  opacity: 0;
  transform: translateY(-20px);
  animation: ${slideInFromTop(20)} 0.3s forwards;
}
.slideInFromBottom {
  opacity: 0;
  transform: translateY(20px);
  animation: ${slideInFromBottom(20)} 0.3s forwards;
}
`
const Main = styled(PileDiv)`
  padding: 0px;
`
const Debug = styled.div`
  position: fixed;
  bottom: 0;
  right: 0;
  color: #3e3832;
  overflow: hidden;
  margin: 0.5rem;
  padding: 0em 0.5em;
  z-index: 99999;
`
export interface TopContainerProps {
  className?: string
  children?: JSX.Element | Array<JSX.Element | null>
  fluid?: boolean
  header?: string | JSX.Element
}

export interface TopContainerSelector {
  serverErrorMessage: string | undefined
  error: any | undefined
  expirationTime: Date | undefined
  highContrast: boolean
  modal: ModalContent | undefined
}

const mapState = (state: State): TopContainerSelector => ({
  serverErrorMessage: state.alert.serverErrorMessage,
  error: state.alert.error,
  expirationTime: state.app.expirationTime,
  highContrast: state.ui.highContrast,
  modal: state.ui.modal
})

export const TopContainer: React.FC<TopContainerProps> = ({
  className, children
}: TopContainerProps): JSX.Element => {
  const {
    serverErrorMessage, error, expirationTime, highContrast, modal
  }: TopContainerSelector = useSelector<State, TopContainerSelector>(mapState)
  const dispatch = useDispatch()
  const { t } = useTranslation()

  const onClear = (): void => {
    dispatch(clientClear())
  }

  const getServerErrorMessage = (): string | undefined => {
    return serverErrorMessage ? t(serverErrorMessage) : undefined
  }

  if (_.isNil(window.onerror)) {
    window.onerror = (msg) => {
      dispatch(clientError({ error: msg }))
    }
  }

  const handleModalClose = (): void => {
    dispatch(closeModal())
  }

  const ErrorFallback = ({ error, resetErrorBoundary }: any) => (
    <Error error={error} resetErrorBoundary={resetErrorBoundary} />
  )

  return (
    <NavHighContrast highContrast={highContrast}>
      <GlobalStyle />
      <ErrorBoundary
        FallbackComponent={ErrorFallback}
        onReset={() => {
          // reset the state of your app so the error doesn't happen again
        }}
      >
        <Header
          highContrast={highContrast}
        />
        <Alert
          message={getServerErrorMessage()}
          error={error}
          onClose={onClear}
        />
        {modal !== undefined && (
          <Modal
            highContrast={highContrast}
            appElement={(document.getElementById('main') || document.body)}
            modal={modal}
            onModalClose={handleModalClose}
          />
        )}
        <Main
          id='main'
          role='main'
          className={classNames(className, { highContrast: highContrast })}
        >
          {children}
        </Main>
        <Debug>
          <SessionMonitor
            expirationTime={expirationTime!}
          />
          <Version />
        </Debug>
      </ErrorBoundary>
    </NavHighContrast>
  )
}

TopContainer.propTypes = {
  className: PT.string,
  children: PT.any,
  fluid: PT.bool,
  header: PT.any
}

export default TopContainer
