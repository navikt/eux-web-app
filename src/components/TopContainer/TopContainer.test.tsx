import { alertClear } from 'actions/alert'
import { render } from '@testing-library/react'
import React from 'react'
import { stageSelector } from 'setupTests'
import { TopContainer, TopContainerProps, TopContainerSelector } from './TopContainer'

jest.mock('react-router-dom', () => ({
  Link: () => (<div className='mock-link' />)
}))

const defaultSelector: TopContainerSelector = {
  bannerMessage: undefined,
  bannerStatus: undefined,
  expirationTime: undefined,
  highContrast: false,
  error: undefined
}

jest.mock('actions/alert', () => ({
  alertClear: jest.fn(),
  alertFailure: jest.fn()
}))

describe('components/TopContainer', () => {
  let wrapper: any
  const initialMockProps: TopContainerProps = {
    header: 'mockHeader',
    title: 'title'
  }

  beforeEach(() => {
    stageSelector(defaultSelector, {})
    wrapper = render(
      <TopContainer {...initialMockProps}>
        <div id='TEST_CHILD' />
      </TopContainer>
    )
  })

  it('Renders', () => {
    expect(wrapper.isEmptyRender()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('Has proper HTML structure', () => {
    expect(wrapper.exists('Header')).toBeTruthy()
    expect(wrapper.exists('Alert')).toBeTruthy()
    expect(wrapper.exists('main')).toBeTruthy()
  })

  it('Compute the client error message', () => {
    (alertClear as jest.Mock).mockReset()
    stageSelector(defaultSelector, { bannerMessage: 'mockMessage|mockParams' })
    wrapper = render(
      <TopContainer {...initialMockProps}>
        <div id='TEST_CHILD' />
      </TopContainer>
    )
    const clientAlert = wrapper.find('Alert[type="server"]')
    expect(clientAlert.render().text()).toEqual('error' + 'mockMessage|mockParams')

    clientAlert.find('Icons').simulate('click')
    expect(alertClear).toHaveBeenCalled()
  })
})
