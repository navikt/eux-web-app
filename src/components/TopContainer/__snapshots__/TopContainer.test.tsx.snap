// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TopContainer Renders 1`] = `
<TopContainer
  header="mockHeader"
>
  <Header
    className=""
  >
    <header
      className="c-header"
    >
      <div
        className="c-header__brand"
      >
        <Link
          className="ml-2 mr-2"
          onClick={[Function]}
          to="/"
        >
          <div
            className="mock-link"
          />
        </Link>
        <div
          className="c-header__skillelinje"
        />
        <div
          className="c-header__tittel"
        >
          <span>
            ui:app-name
          </span>
        </div>
      </div>
      <Undertittel>
        <n
          type="undertittel"
        >
          <h2
            className="typo-undertittel"
          >
            ui:app-title
          </h2>
        </n>
      </Undertittel>
      <div
        className="c-header__saksbehandler"
      >
        <Lenke
          className="c-header__highcontrast-link mr-3"
          href="#highContrast"
          onClick={[Function]}
        >
          <a
            className="lenke c-header__highcontrast-link mr-3"
            href="#highContrast"
            onClick={[Function]}
          >
            ui:label-highContrast
          </a>
        </Lenke>
      </div>
    </header>
  </Header>
  <Alert
    onClose={[Function]}
    type="server"
  />
  <main
    className="_container p-0 container-fluid"
    id="main"
    role="main"
  >
    <div
      id="TEST_CHILD"
    />
  </main>
</TopContainer>
`;
